Move advisory lock to it's own connection

eileencodes opened the pull request:
This PR moves advisory lock to it's own connection instead of
`ActiveRecord::Base` to fix #37748. As a note the issue is present on
both mysql and postgres. We don't see it on sqlite3 because sqlite3
doesn't support advisory locks.

The underlying problem only appears if:

1) the app is using multiple databases, and therefore establishing a new
connection in the abstract models
2) the app has a migration that loads a model (ex `Post.update_all`)
which causes that new connection to get established.

This is because when Rails runs migrations the default connections are
established, the lock is taken out on the `ActiveRecord::Base`
connection. When the migration that calls a model is loaded, a new
connection will be established and the lock will automatically be
released.

When Rails goes to release the lock in the ensure block it will find
that the connection has been closed. Even if the connection wasn't
closed the lock would no longer exist on that connection.

We originally considered checking if the connection was active, but
ultimately that would hide that the advisory locks weren't working
correctly because there'd be no lock to release.

We also considered making the lock more granular - that it only blocked
on each migration individually instead of all the migrations for that
connection. This might be the right move going forward, but right now
multi-db migrations that load models are very broken in Rails 6.0 and
master.

John and I don't love this fix, it requires a bit too much knowledge of
internals and how Rails picks up connections. However, it does fix the
issue, makes the lock more global, and makes the lock more resilient to
changing connections.

Co-authored-by: John Crepezzi <john.crepezzi@gmail.com>

cc/ @rafaelfranca @tenderlove @jhawthorn @matthewd @seejohnrun

eileencodes committed:
Move advisory lock to it's own connection

This PR moves advisory lock to it's own connection instead of
`ActiveRecord::Base` to fix #37748. As a note the issue is present on
both mysql and postgres. We don't see it on sqlite3 because sqlite3
doesn't support advisory locks.

The underlying problem only appears if:

1) the app is using multiple databases, and therefore establishing a new
connetion in the abstract models
2) the app has a migration that loads a model (ex `Post.update_all`)
which causes that new connection to get established.

This is because when Rails runs migrations the default connections are
established, the lock is taken out on the `ActiveRecord::Base`
connection. When the migration that calls a model is loaded, a new
connection will be established and the lock will automatically be
released.

When Rails goes to release the lock in the ensure block it will find
that the connection has been closed. Even if the connection wasn't
closed the lock would no longer exist on that connection.

We originally considered checking if the connection was active, but
ultimately that would hide that the advisory locks weren't working
correctly because there'd be no lock to release.

We also considered making the lock more granular - that it only blocked
on each migration individually instead of all the migrations for that
connection. This might be the right move going forward, but right now
multi-db migrations that load models are very broken in Rails 6.0 and
master.

John and I don't love this fix, it requires a bit too much knowledge of
internals and how Rails picks up connections. However, it does fix the
issue, makes the lock more global, and makes the lock more resilient to
changing connections.

Co-authored-by: John Crepezzi <john.crepezzi@gmail.com>

jhawthorn commented on activerecord/lib/active_record/migration.rb:
```diff
@@ -1374,9 +1374,14 @@ def use_advisory_lock?
 
+      AdvisoryLockBase = Class.new(ActiveRecord::Base) do
+        self.connection_specification_name = "AdvisoryLockBase"
+      end
```
For `SchemaMigration` and `InternalMetadata` we do

``` ruby
def self._internal?
  true
end
```

Feels like we should do the same here

jhawthorn replied:
Also maybe this should be `self.abstract_class = true`?

eileencodes commented:
Ok I moved the AdvisoryLockBase into its own class, added `_internal`, and `abstract_class`. This fixed the test failures in Railties due to framework loading.

matthewd commented:
My grasp of the connection model is slipping (I guess I don't understand how ApplicationRecord's connection-twiddling would affect a connection belonging to its superclass).. but does this literally create an additional TCP connection to the database server? If so, that seems Rather Undesirable.. a database in maintenance mode might only allow a single connection (or a single connection from a maintainer-level user).

IIRC there was a specific reason it was important for the lock to wrap the full set of migrations, but I don't recall what it was -- we might have to find the corresponding PR to check that.

eileencodes commented:
> My grasp of the connection model is slipping (I guess I don't understand how ApplicationRecord's connection-twiddling would affect a connection belonging to its superclass)

A default connection is opened by https://github.com/rails/rails/blob/master/activerecord/lib/active_record/railtie.rb#L204 on `ActiveRecord::Base`. Then the advisory lock code takes out a lock https://github.com/rails/rails/blob/master/activerecord/lib/active_record/migration.rb#L1379. In multi-db we need a `connects_to` in order to connect the replicas in `ApplicationRecord`. When a migration runs that calls a model like `Post` it will load `ApplicationRecord` and call `connects_to`. `connects_to` eventually calls to `establish_connection` in https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L1042. `establish_connection` will look up the pool by the connection_specification_name. It will find that and remove it here https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L1049

Once it's removed it establishes a new connection, which means that we cannot hold onto the lock. Even worse, the way migrations work we need a lock for each connection so if the lock was meant to be global, even if it wasn't throwing an exception due to a closed connection, we're still not protecting all migrations from a concurrent migration.

> but does this literally create an additional TCP connection to the database server

Yea it does, and I don't see any other way around this. I'm open to other ideas but everything we explored was not possible. 

At the moment the only fix for this is saying we do not support advisory locks for multiple databases. We have no way of knowing if an app is multi-db so we can't even check for that and warn the user.

eileencodes commented:
@matthewd do you have any ideas how we can fix the problem? We either need a more global lock (that protects all migrations regardless of connection) or a more granular lock (only protects the current running migration).

matthewd commented:
> I don't understand how ApplicationRecord's connection-twiddling would affect a connection belonging to its superclass

Okay, it turns out my confusion is because that behaviour's special-cased here: https://github.com/rails/rails/blob/2c07d162b392f693ddbba020cad8b23b8b30eb2e/activerecord/lib/active_record/connection_handling.rb#L188

Given that we're re-configuring the same-named connection, I wonder if we could/should recognise that the new connection config matches the already-active one, and therefore we don't need to dis/reconnect at all?

Otherwise I don't think I can suggest anything that wouldn't involve some sort of change in connection management, so if that's off the table, I guess it's just a matter of choosing the less-bad trade-off (which I agree is this PR -- we have to support multi-DB migrations), and apologising to any future reader who's here because migrations started taking two connections. 🤷🏻‍♂️

eileencodes commented:
> Given that we're re-configuring the same-named connection, I wonder if we could/should recognise that the new connection config matches the already-active one, and therefore we don't need to dis/reconnect at all?

So don't remove the connection if it already exists? I did wonder why we do that in the first place. I think it's because you should establish a new connection with a different config but the same name. Eventually maybe that shouldn't be possible. Now that we're using objects everywhere it might be a lot easier for us to inspect the connection db config and figure out if it is the same and if it is the same don't remove it.

I think that's a little more involved and probably an improvement for 6.1. I'm thinking we should merge this for 6.0 and 6.1 because migrations are effectively broken and then work on not removing the connection in 6.1 and undoing this change if we're successful. Does that sound like a fair compromise? I'll add a CHANGELOG as well.

trammel commented:
This fix got merged to master, did it also make it in to the 6.0.2.2 branch/release?

eileencodes commented:
> did it also make it in to the 6.0.2.2 branch/release?

No, 6.0.2.2 was only security and 6.0.2.1 was also only security. This will make it into the 6.0.3 release.
